File: 3604.minimum-time-to-reach-destination-in-directed-graph.py
#
# @lc app=leetcode id=3604 lang=python3
#
# [3604] Minimum Time to Reach Destination in Directed Graph
#

# @lc code=start
from typing import List
import heapq

class Solution:
    def minTime(self, n: int, edges: List[List[int]]) -> int:
        graph = [[] for _ in range(n)]
        for u, v, s, e in edges:
            graph[u].append((v, s, e))
        
        dist = [float('inf')] * n
        dist[0] = 0
        pq = [(0, 0)]  # (time, node)
        
        while pq:
            time, u = heapq.heappop(pq)
            if time > dist[u]:
                continue
            for v, s, e in graph[u]:
                depart = max(time, s)
                if depart <= e:
                    arr = depart + 1
                    if arr < dist[v]:
                        dist[v] = arr
                        heapq.heappush(pq, (arr, v))
        
        return dist[n-1] if dist[n-1] != float('inf') else -1

# @lc code=end

File: 3604.minimum-time-to-reach-destination-in-directed-graph.cpp
//
// @lc app=leetcode id=3604 lang=cpp
//
// [3604] Minimum Time to Reach Destination in Directed Graph
//

// @lc code=start
class Solution {
public:
    int minTime(int n, vector<vector<int>>& edges) {
        vector<vector<array<int, 3>>> adj(n);
        for (auto& edge : edges) {
            adj[edge[0]].push_back({edge[1], edge[2], edge[3]});
        }
        
        vector<long long> dist(n, LLONG_MAX);
        priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<pair<long long, int>>> pq;
        
        dist[0] = 0;
        pq.push({0, 0});
        
        while (!pq.empty()) {
            long long t = pq.top().first;
            int u = pq.top().second;
            pq.pop();
            
            if (t > dist[u]) continue;
            
            if (u == n - 1) return (int)t;
            
            for (auto& e : adj[u]) {
                int v = e[0], start = e[1], end = e[2];
                if (t > end) continue;
                
                long long arrival = (t >= start) ? t + 1 : (long long)start + 1;
                
                if (arrival < dist[v]) {
                    dist[v] = arrival;
                    pq.push({arrival, v});
                }
            }
        }
        
        return -1;
    }
};
// @lc code=end